File: elixir/day2/lib/part1.ex
defmodule Part1 do
  defp parse_input(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.map(&parse_line/1)
  end

  defp parse_line(line) do
    line
    |> String.split()
    |> Enum.map(&String.to_integer/1)
  end

  def is_safe([]), do: false
  def is_safe([_]), do: false

  def is_safe(levels) do
    levels
    |> Enum.chunk_every(2, 1, :discard)
    |> Enum.map(fn [a, b] -> b - a end)
    |> check_differences()
  end

  defp check_differences(diffs) do
    with false <- Enum.any?(diffs, &(abs(&1) < 1 || abs(&1) > 3)) do
      Enum.all?(diffs, &(&1 > 0)) || Enum.all?(diffs, &(&1 < 0))
    else
      true -> false
    end
  end

  defp count_safe_reports(reports) do
    Enum.count(reports, &is_safe/1)
  end

  def part1(input) do
    input
    |> parse_input()
    |> count_safe_reports()
  end
end

File.read!(~c"input.txt")
|> Part1.part1()
|> IO.inspect(label: "part1")

File: ruby/day2/part1.rb
# frozen_string_literal: true

def parse_input(input)
  input.split("\n").map { |line| line.split.map(&:to_i) }
end

def count_safe_reports(reports)
  reports.select { safe?(_1) }.count
end

def safe?(report)
  return false if report.length < 2

  diffs = report.each_cons(2).map { |a, b| a - b }
  return false unless diffs.all? { |d| d.abs.between?(1, 3) }

  diffs.all?(&:positive?) || diffs.all?(&:negative?)
end

def part1(input)
  reports = parse_input(input)
  count_safe_reports(reports)
end

if __FILE__ == $PROGRAM_NAME
  input = File.read('input.txt')
  p part1(input)
end
